back-end/helpers/get/transcript_data_grouper.js

```diff
@@ -0,0 +1,159 @@
+                try {
+                    await this.#mergeIntent(comparedIntent);
+                    return true;
+                } catch (e) {
+                    console.log(e.message)
+                    return false;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @param Object Formatted intent to compare to
+     * @return number in [0, 1]
+     */
+    #generateSimilarityScore(comparedIntent) {
+        const splitIntent1 = this.intent.question.split(" ");
+        const splitIntent2 = comparedIntent.question.split(" ");
+        this.differentWordsIndices = [];
+
+        // If the length is significantly different we do not want to merge
+        if (Math.abs(splitIntent1.length - splitIntent2.length) > 3) {
+            return 0;
+        }
+
+        // Set a current score and let it decrement by 1 * weight each time a word is different
+        const DECREMENTOR = 1;
+        let score = splitIntent1.length;
+
+        // Set the weight of how much each difference matters
+        let weight = 1;
+        if (splitIntent1.length < 10) {
+            weight = 0.5;
+        }
+
+        // We use index to give us a greater control of manipulation over both arrays
+        let index1 = 0;
+        let index2 = 0;
+
+        while (index1 < splitIntent1.length && index2 < splitIntent2.length) {
+            if (splitIntent1[index1] !== splitIntent2[index2] && splitIntent2 !== "[merged]") {
+                score -= weight * DECREMENTOR;
+                this.differentWordsIndices.push(index1);
+
+                // Attempt to not be off-by-1 index for names by identifying titles
+                if (titles.includes(splitIntent1[index1].toUpperCase())) {
+                    index1 += 1;
+                }
+                if (titles.includes(splitIntent2[index2].toUpperCase())) {
+                    index2 += 1;
+                }
+            }
+            index1 += 1;
+            index2 += 1;
+        }
+        return score / splitIntent1.length;
+    }
+
+    /**
+     * @param Object Formatted intent to merge to
+     * Merges 2 intents
+     */
+    async #mergeIntent(comparedIntent) {
+        if (comparedIntent.previousIntents) {
+            // Make deep copy of original intent to not create recursion
+            let original = JSON.parse(JSON.stringify(comparedIntent.previousIntents[0]));
+            // Calculate percentages of each intent previously merged with the original
+            for (let i = 1; i < comparedIntent.previousIntents.length; i++) {
+                await this.#recalculatePercentages(comparedIntent.previousIntents[i], original);
+            }
+            original.previousIntents = comparedIntent.previousIntents;
+            comparedIntent = original;
+        } else {
+            comparedIntent.previousIntents = [];
+            const copyComparedIntent = JSON.parse(JSON.stringify(comparedIntent))
```

Comment:
Doesn't previousIntents just need to be the question? Why are we copying the whole object. If it is an issue of being able to recalculate percentages, i think it would be more efficient to create a new object with the most important parts of the intent

Reply:
So the reason why we have to put in the whole intent is because later in the recalculate percentage it expects the passed in thing to be an object with field question

Reply:
I think it would be more efficient to create an object that fits the specs than create a deep copy